fendermint/vm/interpreter/src/fvm/upgrade_scheduler.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::bail;
+use fendermint_vm_core::chainid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_shared::chainid::ChainID;
+
+use super::state::{snapshot::BlockHeight, FvmExecState};
+
+/// a function type for migration. It takes in a state and returns a new application version (or 0 if not affected)
+// TODO: Add missing parameters
+// TODO: Create a application version type
+pub type MigrationFunc<DB> = fn(state: &mut FvmExecState<DB>) -> anyhow::Result<u64>;
+
+/// Upgrade represents a single upgrade to be executed at a given height
+#[derive(Clone)]
+pub struct Upgrade<DB>
+where
+    DB: Blockstore + 'static + Clone,
+{
+    /// the chain id on which the upgrade should be executed
+    chain_id: ChainID,
+    /// the block height at which the upgrade should be executed
+    block_height: BlockHeight,
+    /// the migration function to be executed
+    migration: MigrationFunc<DB>,
+
+    /// the chain name is never read after initialization
+    _chain_name: String,
+}
+
+impl<DB> Upgrade<DB>
+where
+    DB: Blockstore + 'static + Clone,
+{
+    pub fn new(
+        chain_name: String,
+        block_height: BlockHeight,
+        migration: MigrationFunc<DB>,
+    ) -> anyhow::Result<Self> {
+        let chain_id = chainid::from_str_hashed(&chain_name)?;
+
+        let me = Self {
+            chain_id,
+            block_height,
+            migration,
+            _chain_name: chain_name,
+        };
+
+        Ok(me)
+    }
+
+    pub fn execute(&self, state: &mut FvmExecState<DB>) -> anyhow::Result<u64> {
+        (self.migration)(state)
+    }
+}
+
+/// UpgradeScheduler represents a list of upgrades to be executed at given heights
+/// During each block height we check if there is an upgrade scheduled at that
+/// height, and if so the migration for that upgrade is performed.
+#[derive(Clone)]
+pub struct UpgradeScheduler<DB>
+where
+    DB: Blockstore + 'static + Clone,
+{
+    // TODO: Consider using Set
```

Comment:
Why not use a `BTreeMap` which is ordered by chain ID and block height?

Reply:
@aakoshh yeah, had it BTreeMap originally, but noticed that `ChainID` did not implement the `std::cmp::Ord` trait so left a todo instead.

But will add it in now, I think it should work to add a custom key type which implements the Ord trait that allows me to call `Into` on the ChainID